Update the undocumented graphical python examples

vinnik-dmitry07 opened the pull request.

vinnik-dmitry07 committed:
Fix k-d tree link.

(cherry picked from commit 49194d0bd0c266b402566aec3eb3dbf5d945ae2b)

vinnik-dmitry07 committed:
Fix formula in KernelDensity description.

(cherry picked from commit e7d67d9fbe1072edb68ddaf254e2190dbf10ce5c)

vinnik-dmitry07 committed:
Remove RealFeatures from util.py, mclda.py.

(cherry picked from commit f8a0a05b752950bc3588e23b20c806b981de642c)

vinnik-dmitry07 committed:
Update matplotlib import in mclda.py.

(cherry picked from commit 3da682000e4aae226766c0fa906eb9e09a9df09c)

vinnik-dmitry07 committed:
Minor change in mclda.py: reformat.

(cherry picked from commit 56fccb7b23a55c836e47e6d35687d79df256c3ad)

vinnik-dmitry07 committed:
Update matplotlib calls in lda.py.

(cherry picked from commit 39fbbb55fe9d5bc05571862143352c4668333892)

vinnik-dmitry07 committed:
Minor changes to lda.py: reformat.

(cherry picked from commit 1814ebdac9f079ccee9424935786f523724d511c)

gf712 commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
 neg = np.array([x_neg, y_neg])
-features = RealFeatures( np.array(np.concatenate([pos, neg], 1)) )
+
+features_ = features(np.array(np.concatenate([pos, neg], 1)))
```
@karlnapf maybe we should stop `from shogun import *` to avoid this kind of thing happening?

karlnapf replied:
yes absolutely, we should use `import shogun as sg` and then `sg.features` (actually also in meta examples, Ill open an issue)

karlnapf replied:
#4990

karlnapf commented on examples/undocumented/python/graphical/lda.py:
```diff
@@ -1,32 +1,31 @@
-from pylab import figure,pcolor,scatter,contour,colorbar,show,subplot,plot,connect
+import matplotlib.pyplot as plt
```
I would put changes to python graphical example in a different PR than the C++ code below. EDIT. sorry I only now realised that those are cleanups....fine to have a single PR then

Also, note that graphical python examples are pretty much the bottom of our priority list, so let's get these changes in but let's not do more of these. Instead, I suggest you focus on sending some C++ code

vinnik-dmitry07 replied:
I made these changes for vigsterkr can run this example, because there, and I guessed in some other places, was also the problem with `const std::shared_ptr<DenseFeatures<float64_t>>&`. I mean I did this, not because of the priorities. But now I have understood that it is a SWIG problem.

karlnapf replied:
sure thing. the patch is welcome, let's get it in, just wanted to make sure you are on the same page as us

karlnapf replied:
and it does make sense to improve the whole listing and make it ready for the new api, so then we don't have to touch it again anytime soon.

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
+import matplotlib.pyplot as plt
 import numpy as np
-import util
+from shogun import features
```
actually I just looked through the changes here, the only functional change is replacing RealFeatures with Features.
The file above has no functional changes as far as I can see.

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
-from shogun import RealFeatures
 from shogun import MulticlassLabels
```
could we have a `import shogun as sg` as per the issue I referenced?

karlnapf replied:
(and remove all the other imports?)

vinnik-dmitry07 replied:
Aye

karlnapf commented on examples/undocumented/python/graphical/util.py:
```diff
@@ -1,10 +1,14 @@
 import pylab
-from numpy import ones, array, double, meshgrid, reshape, linspace, \
-	concatenate, ravel, pi, sinc
+from numpy import ones, array, double, meshgrid, linspace, concatenate, ravel, pi, sinc
```
import numpy as np

karlnapf commented on src/shogun/distributions/KernelDensity.h:
```diff
@@ -50,8 +50,8 @@ enum EEvaluationMode
  * way to estimate an unknown pdf. The pdf at a query point given finite training samples is calculated using the
- * following formula : \\
- * \f$pdf(x')= \frac{1}{nh} \sum_{i=1}^n K(\frac{||x-x_i||}{h})\f$ \\
+ * following formula: \n
```
I think there are centred math tex commands in doxygen, so you dont have to do newlines by hand. See e.g. GaussianKernel.h
```
 * \f[
 * k({\bf x},{\bf x'})= exp(-\frac{||{\bf x}-{\bf x'}||^2}{\tau})
 * \f]
```

karlnapf commented on src/shogun/multiclass/tree/KDTree.h:
```diff
@@ -40,7 +40,7 @@ namespace shogun
 
 /** @brief This class implements KD-Tree.
- * cf. http://www.autonlab.org/autonweb/14665/version/2/part/5/data/moore-tutorial.pdf
+ * cf. https://web.archive.org/web/20140327060654if_/http://www.autonlab.org:80/autonweb/14665/version/2/part/5/data/moore-tutorial.pdf
```
is this really a good idea? Why not put a reference as a citation (a different/better reference, or wikipedia)

vinnik-dmitry07 replied:
I am not the author of code, so I decided to put what the author meant when was writing this comment

vinnik-dmitry07 replied:
I saw that link from the archive works, I do not really investigate content. Sorry, will do better next time.

karlnapf replied:
no worries at all. I just think dead-links should be removed rather than pulled out from the internet archive ;) wikepedia is probably fine here anyways.

karlnapf commented on examples/undocumented/python/graphical/lda.py:
```diff
@@ -1,32 +1,31 @@
 from shogun import *
 import util
 
-util.set_title('LDA')
```
for the record. next time please make a smaller PR title. This one is way too epic ;)

karlnapf commented on examples/undocumented/python/graphical/lda.py:
```diff
@@ -1,32 +1,31 @@
-features=util.get_realfeatures(pos, neg)
-lda=LDA(gamma, features, labels)
+labels = util.get_labels()
+features = util.get_realfeatures(pos, neg)
```
while we are at it `get_real_features`

karlnapf commented on examples/undocumented/python/graphical/lda.py:
```diff
@@ -1,32 +1,31 @@
-lda=LDA(gamma, features, labels)
+labels = util.get_labels()
+features = util.get_realfeatures(pos, neg)
+lda = LDA(gamma, features, labels)
```
white we are at it. Could you port this to the new API usage, something in the lines of
```
lda = machine("LDA", gamma=gamma, labels=labels)
lda.train(features)
```

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
 
 # train qda
-labels = MulticlassLabels( np.concatenate([np.zeros(N), np.ones(N)]) )
+labels = MulticlassLabels(np.concatenate([np.zeros(N), np.ones(N)]))
```
use `sg.labels` factory (new api)

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
+
+features_ = features(np.array(np.concatenate([pos, neg], 1)))
 
 lda = MCLDA()
```
new API

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
+features_ = features(np.array(np.concatenate([pos, neg], 1)))
 
 lda = MCLDA()
 lda.set_labels(labels)
```
use `put` (or rather a kwarg when using the `machine` factory)

karlnapf commented on examples/undocumented/python/graphical/mclda.py:
```diff
@@ -1,63 +1,57 @@
 
-dense = RealFeatures( np.array((np.ravel(x), np.ravel(y))) )
+dense = features(np.array((np.ravel(x), np.ravel(y))))
 dense_labels = lda.apply(dense).get_labels()
```
use `get`

karlnapf replied:
`lda.apply(dense).get('labels')`

vinnik-dmitry07 committed:
Change "from x import .." to "import x as .." where x is numpy, shogun.

(cherry picked from commit 2ad1148e7284731efd8e9c6ed151af50f80d601c)

vinnik-dmitry07 committed:
Correct matplotlib, numpy, shogun import in classifier_gaussian_process_binary_classification.py. Remove RealFeatures.

(cherry picked from commit 597ddd1c1c1b90da97a65397f4352e9e1d7b14f4)

vinnik-dmitry07 committed:
Minor changes to classifier_gaussian_process_binary_classification.py: reformat.

(cherry picked from commit 06ec2b35e6b42d511ec7a2f256312f4fe5208e5b)

vinnik-dmitry07 committed:
Fully fix all commit files.

(cherry picked from commit 96ef0450e4a5d2d587e70b5c1c19ab436da684b0)

karlnapf commented on examples/undocumented/python/graphical/classifier_gaussian_process_binary_classification.py:
```diff
@@ -1,61 +1,52 @@
-    train_features = RealFeatures(X_train)
-    train_labels = BinaryLabels(y_train)
+    train_features = sg.features(X_train)
+    train_labels = sg.BinaryLabels(y_train)
```
sg.labels should work

karlnapf commented on examples/undocumented/python/graphical/classifier_gaussian_process_binary_classification.py:
```diff
@@ -1,61 +1,52 @@
 
     # create zero mean function
-    mean = ZeroMean()
+    mean = sg.ZeroMean()
```
is there a factory for mean functions? I think so as we have a gp meta example

vinnik-dmitry07 replied:
I did not find.

karlnapf commented on examples/undocumented/python/graphical/classifier_gaussian_process_binary_classification.py:
```diff
@@ -73,35 +64,35 @@ def gaussian_process_binary_classification_laplace(X_train, y_train, n_test=50):
 
     # create and train GP classifier, which uses Laplace approximation
-    gp = GaussianProcessClassification(inf)
+    gp = sg.GaussianProcessClassification(inf)
```
all those should have a factory, meta examples should guide you

vinnik-dmitry07 replied:
![image](https://user-images.githubusercontent.com/24797359/78395822-e4d83280-75f6-11ea-9b63-73dca4ffad3c.png)

vinnik-dmitry07 replied:
How to run this method from the machine? (get_probabilities)

karlnapf replied:
just checked, it is not possible currently, and we need to refactor this stuff (meta example is also not yet ported to new api I realised). Ignore my comment then

karlnapf replied:
#4996

vinnik-dmitry07 committed:
Fix link 2

(cherry picked from commit faa83c212104695531a8c9b9362864746138e511)

karlnapf reviewed (commented):
Thanks for the update.
I'd like to encourage you (again) to focus on c++ code though. These graphical example are not super important. But since you have made the effort in polishing them, let's finish the job (on those you touched), but please don't add new ones to the list for now.
I made a few comments, mostly regarding factories. Try to use them where you can. If none exist, open an issue for it (you don't have to do it here)

Finally, I think we should add those examples to the CI (with a fake matplotlib), so that they get at least executed. Thoughts @vigsterkr ?

vinnik-dmitry07 commented:
> I'd like to encourage you (again) to focus on c++ code though
> but please don't add new ones to the list for now

Okidoki. Btw these examples look really nice, should be in the showroom, but it does not work :(.
![image](https://user-images.githubusercontent.com/24797359/78391275-bbb3a400-75ee-11ea-87f3-e88c75da3c07.png)

vinnik-dmitry07 commented:
Also here is a problem with SPE_LOCAL (not caused by changes).

karlnapf commented:
We could add those to notebooks for example, but let's do that later.
Any idea what the problem with SPE is?

vinnik-dmitry07 committed:
Rename m_nupdates to m_num_updates. Changed the tolerance type: float32_t -> float64_t. Fixed get_tolerance return type.

(cherry picked from commit b2b1332d83dbecf825c22c2f87df153f1776e460)

vinnik-dmitry07 committed:
Fixed the StochasticProximityEmbedding typos.

(cherry picked from commit 23fda5cbe3d931b1f90f6eb069f73834c5fdff78)

vinnik-dmitry07 committed:
Reformat StochasticProximityEmbedding.

(cherry picked from commit 82e6083e677d89e184b5a863c2a0e651e8f99600)

vinnik-dmitry07 committed:
The changes I am not sure you be glad to see =)

(cherry picked from commit 89bb177afbfdffc195f5bb1cd43cb213626b5e9a)

vinnik-dmitry07 committed:
sg.BinaryLabels -> sg.labels. Fix spe_helix example.

(cherry picked from commit 5ba07399592fa28e4e98fec8e3bfd28134741725)

vinnik-dmitry07 commented:
> We could add those to notebooks for example, but let's do that later.
> Any idea what the problem with SPE is?

I finally figured out that m_k should be about 400 (from total 500 points), also along the way I fixed some bug and added missed parameter.
![image](https://user-images.githubusercontent.com/24797359/78403072-6d110480-7604-11ea-9755-a0a0f0fee025.png)

karlnapf commented on examples/undocumented/python/graphical/classifier_perceptron_graphical.py:
```diff
@@ -1,78 +1,75 @@
-	# 2D data
-	_DIM = 2
+    # To get the nice message that the perceptron has converged
+    # dummy = sg.BinaryLabels()
```
this line is unnecessary or?
I'd also remove the comment about the info loglevel

vinnik-dmitry07 replied:
Yes, Perceptron now does not use SG_INFO, I will remove it.

karlnapf commented on examples/undocumented/python/graphical/converter_ffsep_bss.py:
```diff
@@ -7,12 +7,9 @@
 Kevin Hughes 2013
```
question: could we group all the ica examples into one? The code is more or less the same. We could just have a loop over all the different ica methods

karlnapf commented on examples/undocumented/python/graphical/util.py:
```diff
@@ -1,115 +1,117 @@
+        return data
+    else:
+        if type == 'binary':
+            return sg.BinaryLabels(data)
```
`labels` factory pls

karlnapf replied:
and then you shouldn't need the if then else, as the factory does that

karlnapf commented on src/shogun/converter/StochasticProximityEmbedding.cpp:
```diff
@@ -29,19 +28,19 @@ void StochasticProximityEmbedding::init()
 	SG_ADD(&m_tolerance, "m_tolerance", "Regularization parameter");
-	SG_ADD(&m_max_iteration, "max_iteration", "maximum number of iterations");
+	SG_ADD(&m_num_updates, "m_num_updates", "SPE number of updates");
+	SG_ADD(&m_max_iteration, "max_iteration", "Maximum number of iterations");
```
@gf712 are we still using SG_ADD; or is the templated watch already there?

gf712 replied:
The PR is mostly ready I think. I can have a look tomorrow!

karlnapf commented on src/shogun/converter/StochasticProximityEmbedding.h:
```diff
@@ -135,39 +137,36 @@ class StochasticProximityEmbedding : public EmbeddingConverter
-		float32_t m_tolerance;
+		float64_t m_tolerance;
 
-		/** number of apdates per SPE iteration */
```
"apdates" :laughing:

karlnapf reviewed (commented):
looks good, few more minor fixes and then we can merge it

vinnik-dmitry07 committed:
Minor changes + update kernel_ridge_regression.py, kernel_ridge_regression_sinc.py

(cherry picked from commit 876f24c955f3636d7cf338f2a335387c9f287eaa)

vinnik-dmitry07 committed:
Compound the examples of converter algorithms.

(cherry picked from commit d3e5b9147716328bcf9a200b855924bfe45ff055)

vinnik-dmitry07 commented:
Sorry, I remember that I do not have to touch more files but to test nothing breaks after changing this: https://github.com/shogun-toolbox/shogun/pull/4989#discussion_r403453180 I had to update kernel_ridge_regression.

vinnik-dmitry07 commented:
Accidentally-closed

vinnik-dmitry07 committed:
Minor changes.

(cherry picked from commit ab53d98150a4b5b8b623824a84b314e437ebca70)

vinnik-dmitry07 committed:
Fix the transfer from BinaryLabels.

(cherry picked from commit 65e3ca23e0ca7dd50fb36e90cbc894edf6a47846)

karlnapf commented on examples/undocumented/python/graphical/kernel_ridge_regression_sinc.py:
```diff
@@ -1,31 +1,30 @@
+lab = sg.labels(Y.flatten())
+gk = sg.kernel('GaussianKernel', log_width=width)
+gk.init(feat, feat)
+krr = sg.machine('KernelRidgeRegression')
```
could you use python `**kwargs`, i.e.
`krr = sg.machine('KernelRidgeRregression", labels=lab, kernel=gk, tau=1e-3)` (and for all the other instantiations as well). Makes everything look more like Python :)

karlnapf commented on examples/undocumented/python/graphical/kernel_ridge_regression_sinc.py:
```diff
@@ -1,31 +1,30 @@
-show()
+plt.plot(XE[0], YE, label='test output')
+plt.plot([XE[0, 200]], [YE200], '+')
+# print(YE[200], YE200)
```
remove

karlnapf commented on src/shogun/converter/StochasticProximityEmbedding.h:
```diff
@@ -7,72 +7,74 @@
 
 		/** destructor */
-		virtual ~StochasticProximityEmbedding();
+		~StochasticProximityEmbedding() override;
```
no virtual destructor?

vinnik-dmitry07 replied:
I follow https://en.cppreference.com/w/cpp/language/virtual#In_detail

vinnik-dmitry07 replied:
More here https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html

gf712 replied:
Hmm but this is a destructor. You are overriding the base class destructor so there will be leaks..

vinnik-dmitry07 replied:
"override" keyword simply means "this function is marked as virtual in some base class"

vinnik-dmitry07 replied:
Concluding everything, all StochasticProximityEmbedding base classes till SGObject (EmbeddingConverter, Converter, Transformer) should have the override keyword instead of virtual.

vinnik-dmitry07 replied:
Imagine we have a base class B having a method X and derived classes D_1, D_2, ..., D_N declared so: class D_1 : B; class D_2: D_1; ... 
The most correct way is to declare X as virtual in B and as override it in D_1, ..., D_N-1 and as final in D_N.

gf712 replied:
Just had a look online. This override doesn’t do anything. The destructor is always virtual, because it needs to lookup the vtable along the hierarchy to know how to destruct all class members. The override only ensure that the base class is virtual, which is a given in the way shogun is designed from SGObject down. Imo this syntax is confusing and doesn’t bring anything. It says override, but it doesn’t, and to me at lest it looks like a bug (even though it actually isn’t).

gf712 replied:
Btw I know how virtual functions work. My point was that overriding a destructor imo would imply overriding the behaviour of the parent classes. It doesn’t, because the destructor override here is a noop.

gf712 replied:
https://github.com/isocpp/CppCoreGuidelines/issues/721

vinnik-dmitry07 replied:
> It says override, but it doesn’t

[C++11 Standart](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf)
"Even though destructors are not inherited, a destructor in a derived class **_overrides_** a base class destructor." -- Page 233.
<br>
"The use of the _virtual_ specifier in the declaration of an overriding function is legal but redundant (has empty semantics)." -- Page 231.

gf712 replied:
But in that case what's the point of `override`, if it overriden anyway? Other than to make sure that the base class has a virtual destructor, which we know it does? Anyway. up to you, my point is that then we need to change every destructor in shogun, otherwise we just end up with a mix and match of patterns and code styles.

vinnik-dmitry07 replied:
> Anyway. up to you, my point is that then we need to change every destructor in shogun

I can do it in an hour without any future problems.

>But in that case what's the point of override, if it overriden anyway?

C++ headers were invented to developers who use someone else code. When you first look at a header file, you do not know anything about the base class. As using _virtual_ specifier in the derived class has no sense, we should specify somehow that the method overrides the base method. Plus using override by analogy to writing _private_ specifier helps us not to shoot ourselves in the foot.

vinnik-dmitry07 replied:
https://github.com/chromium/chromium/search?q=override&unscoped_q=override
Chromium guys agree with me.

gf712 replied:
go for it if you think that is worth it :)

gf712 replied:
for reference I think @vinnik-dmitry07 means this https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/tools/clang/plugins/tests/virtual_specifiers.txt

karlnapf replied:
@vinnik-dmitry07 could you hold your horses for now and just follow shogun's style? We don't want to mix up ways we do things, as this leads to redundant discussions and compiler warnings. If you are keen in changing the way this is done in shogun, please open an issue and a separate PR that applies what you want to do to all of shogun.

karlnapf replied:
Of course such a PR is extremely welcome. (Although I suggest you defer this to later and focus on sending something related to your project proposal for now)

karlnapf commented on src/shogun/converter/StochasticProximityEmbedding.h:
```diff
@@ -102,31 +104,31 @@ class StochasticProximityEmbedding : public EmbeddingConverter
 		 * @param max_iteration the maximum number of iterations
 		 */
-		void set_max_iteration(const int32_t max_iteration);
+		void set_max_iteration(int32_t max_iteration);
```
++

gf712 replied:
Was the const it causing issues?

vinnik-dmitry07 replied:
Clang-tidy says: https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html

gf712 replied:
I see. OK!

karlnapf commented on src/shogun/converter/StochasticProximityEmbedding.h:
```diff
@@ -135,39 +137,36 @@ class StochasticProximityEmbedding : public EmbeddingConverter
 
 		/** get name */
-		virtual const char* get_name() const;
+		const char* get_name() const override;
```
I am actually not sure what our policy here is, we used to keep the `virtual`. But it is not needed with `override` I guess.
@gf712 @vigsterkr thoughts?

gf712 replied:
This is the correct way to do it, but I think it causes lots of compiler warnings... because we always use virtual... We should write a libtooling script to replace all the virtual get_name with override, or even just use sed

gf712 replied:
@vinnik-dmitry07 whilst you are at it can you also fix all of these, please?

vinnik-dmitry07 committed:
Replace put with kwargs.

(cherry picked from commit a2d57fe427fd9a9543023e2182bd553beed1e601)

vinnik-dmitry07 committed:
Correct the formula 2.

(cherry picked from commit 75dcfc419f792c1670f9f01b5a895a67853e4e21)

vinnik-dmitry07 committed:
Revert changing to override and default.

karlnapf reviewed (approved):
LGTM
thanks a lot for the tidy up!

karlnapf commented:
somebody should eyeball this and then we can merge it

gf712 commented:
To be on the safe side can you trigger CI, please? Now all the builds are working again!

vinnik-dmitry07 committed:
Trigger CI.
